gio/src/subclass/action.rs

```diff
@@ -0,0 +1,427 @@
+        iface.change_state = Some(action_change_state::<T>);
+        iface.activate = Some(action_activate::<T>);
+
+        unsafe {
+            let class_ref = glib::object::Class::<glib::Object>::from_type(instance_type).unwrap();
+            let object_class =
+                class_ref.as_ref() as *const _ as *mut glib::gobject_ffi::GObjectClass;
+
+            let mut first_prop = std::mem::MaybeUninit::uninit();
+            let properties = glib::gobject_ffi::g_object_class_list_properties(
+                object_class,
+                first_prop.as_mut_ptr(),
+            );
+            glib::ffi::g_free(properties as *mut _);
+            let first_prop = first_prop.assume_init() + 1;
+
+            set_first_prop(instance_type, first_prop as usize);
+
+            glib::gobject_ffi::g_object_class_override_property(
+                object_class,
+                first_prop + ActionProperty::Name as u32,
+                b"name\0".as_ptr() as *const _,
+            );
+            glib::gobject_ffi::g_object_class_override_property(
+                object_class,
+                first_prop + ActionProperty::ParameterType as u32,
+                b"parameter-type\0".as_ptr() as *const _,
+            );
+            glib::gobject_ffi::g_object_class_override_property(
+                object_class,
+                first_prop + ActionProperty::StateType as u32,
+                b"state-type\0".as_ptr() as *const _,
+            );
+            glib::gobject_ffi::g_object_class_override_property(
+                object_class,
+                first_prop + ActionProperty::Enabled as u32,
+                b"enabled\0".as_ptr() as *const _,
+            );
+            glib::gobject_ffi::g_object_class_override_property(
+                object_class,
+                first_prop + ActionProperty::State as u32,
+                b"state\0".as_ptr() as *const _,
+            );
+        }
+    }
+}
+
+unsafe extern "C" fn action_get_name<T: ActionImpl>(
+    actionptr: *mut ffi::GAction,
+) -> *const libc::c_char {
+    let instance = &*(actionptr as *mut T::Instance);
+    let imp = instance.imp();
+
+    let instance = imp.obj();
+    let name_quark = {
+        static QUARK: OnceLock<glib::Quark> = OnceLock::new();
+        *QUARK.get_or_init(|| glib::Quark::from_str("gtk-rs-subclass-action-get-name"))
+    };
+    if let Some(old_name_ptr) = instance.qdata::<GString>(name_quark) {
+        old_name_ptr.as_ref().as_ptr()
+    } else {
+        instance.set_qdata(name_quark, imp.name());
+        instance
+            .qdata::<GString>(name_quark)
+            .unwrap()
+            .as_ref()
+            .as_ptr()
+    }
+}
+
+unsafe extern "C" fn action_get_parameter_type<T: ActionImpl>(
+    actionptr: *mut ffi::GAction,
+) -> *const GVariantType {
+    let instance = &*(actionptr as *mut T::Instance);
+    let imp = instance.imp();
+
+    if let Some(parameter_type) = imp.parameter_type() {
```

Comment:
This should probably be moved into the `set_qdata()` call. Now you'd get inconsistent behaviour if the function returns `Some` on the first call and `None` afterwards

Reply:
It's not directly forbidden by Gio's docs, but I think it could be insane if the type of a parameter would change during the action's lifetime.

A similar assumption is made in [`ActionGroup`](https://github.com/gtk-rs/gtk-rs-core/blob/main/gio/src/subclass/action_group.rs#L433).

Reply:
My personal approach here would be to either only call the function once to don't give the user a chance to do it wrong, or otherwise at least to add an assertion :)

Reply:
I've changed this and other places (`name`, `parameter_type` and `state_type`) in the way that the function is called once.